### PR TITLE
Fix unsound optimization with explicit variant discriminants

### LINKED ISSUE #89485: SimplifyBranchSame assumes discriminant is the same as variant index.
```console
$ cat a.rs 
#[derive(Debug, Eq, PartialEq)]
pub enum Type {
    A = 1,
    B = 2,
}
pub fn encode(v: Type) -> Type {
    match v {
        Type::A => Type::B,
        _ => v,
    }
}
fn main() {
  assert_eq!(Type::B, encode(Type::A));
}
$ rustc -Zmir-opt-level=0 a.rs && ./a
$ rustc a.rs && ./a
thread 'main' panicked at 'assertion failed: `(left == right)`
  left: `B`,
 right: `A`', a.rs:13:3
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

Originally found by @bonega  when experimenting with discriminants starting from 1 in https://github.com/rust-lang/rust/pull/88984#issuecomment-932818833.

@rustbot modify labels: +regression-from-stable-to-stable +I-unsound

### PR DESCRIPTION
**Pull request by FabianWolff**
Fixes #89485.

### CODE REVIEW

**Comment by oli-obk** [comment]
lgtm, tho we should do some cleanups of the opt after this is merged

r? @tmiasko for another pair of eyes

**Comment by tmiasko** [comment]
The change seems reasonable.

At the same time I can't really follow the preexisting implementation. Why the left hand side of the assignment is ignored for the purposes of the comparison? How do we know that place that we read the discriminant from still holds the same value?

File: compiler/rustc_mir_transform/src/simplify_try.rs [hunk 1]
```diff
@@ -742,18 +752,18 @@ impl<'a, 'tcx> SimplifyBranchSameOptimizationFinder<'a, 'tcx> {
...
+            if y_target_and_value.value.is_some() => {
                 // choose basic block of x, as that has the assign
-                helper(rhs, place, variant_index, x_target_and_value.target)
+                helper(rhs, place, variant_index, y_target_and_value.value.unwrap(), x_target_and_value.target)
             }
             (
                 StatementKind::SetDiscriminant { place, variant_index },
                 StatementKind::Assign(box (_, rhs)),
             )
             // we need to make sure that the switch value that targets the bb with SetDiscriminant (x), is the same as the variant index
```
Comment (camelid) [inline compiler/rustc_mir_transform/src/simplify_try.rs:763]: nit: Seems like this comment needs to be updated from this change?

File: compiler/rustc_mir_transform/src/simplify_try.rs [hunk 2]
```diff
@@ -750,20 +752,28 @@ impl<'a, 'tcx> SimplifyBranchSameOptimizationFinder<'a, 'tcx> {
...
+            ) if x_target_and_value.value.is_some() => {
                 // choose basic block of y, as that has the assign
-                helper(rhs, place, variant_index, x_target_and_value.value.unwrap(), y_target_and_value.target)
+                helper(
+                    rhs,
+                    place,
+                    variant_index,
+                    x_target_and_value.value.unwrap(),
+                    y_target_and_value.target,
+                )
```
Comment (camelid) [inline compiler/rustc_mir_transform/src/simplify_try.rs]: The formatting seems to have changed in a weird way here. Can you try running `./x.py fmt`?

**Comment by oli-obk** [comment]
@tmiasko so I guess we disable the opt in addition to this fix until we have a clearer picture of how it works?

File: compiler/rustc_mir_transform/src/simplify_try.rs (hunk 2 above)
Comment (camelid) [reply compiler/rustc_mir_transform/src/simplify_try.rs]: Ah, never mind, I didn't realize this got turned into a `match`.

**Comment by tmiasko** [comment]
>  let's disable the opt in addition to this fix until we have a clearer picture of how it works?

Sounds good to me.

**Comment by oli-obk** [comment]
@FabianWolff please also add an early abort in the mir opt with

```
if !tcx.sess.opts.debugging_opts.unsound_mir_opts {
    return;
}
```

so that we don't run the opt at all in normal mode

**Comment by FabianWolff** [comment]
@oli-obk Done.

**Comment by wesleywiser** [comment]
@bors r=oli-obk p=1 

Fixes P-critical issue.

**Comment by wesleywiser** [comment]
Beta nominating so we can get this fix out as fast as possible. The most important change is that we've disabled the pass by default which is safe and should be very easy to backport.

**Comment by nbdd0121** [comment]
Should this be backported to stable?

File: src/test/ui/mir/issue-89485.rs [hunk 3]
```diff
@@ -0,0 +1,18 @@
+// Regression test for issue #89485.
```
Comment (Aaron1011) [inline src/test/ui/mir/issue-89485.rs:1]: I think it would be a good idea to add a copy of this test with `-Zunsound-mir-opts` enabled

**Comment by wesleywiser** [comment]
> Should this be backported to stable?

It's a good question @nbdd0121 . IMO, since we're about 1.5 weeks from the next release, beta backport seems sufficient to me.

**Comment by oli-obk** [comment]
@rustbot label: +perf-regression-triaged

we disabled an unsound MIR opt. LLVM can't figure this out on its own sadly.

**Comment by apiraino** [comment]
Beta backport accepted as per compiler team [on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202021-10-07.20.2354818/near/256585153)

Note from the meeting: "if there’s any issue at all, its also fine to *just* backport solely the bit that disables the pass"

@rustbot label +beta-accepted

**Pull request merged**